### DISCUSSION TITLE
automatically resize plotItem viewBox to show all other viewBoxes

### DISCUSSION THREAD

**Discussion by mesa-git** [pyqtgraph/pyqtgraph#2288] [Q&A] [answered]
cross-posting from stackoverflow: https://stackoverflow.com/questions/72110642/pyqtgraph-resize-plotitem-viewbox-to-show-all-other-viewboxes

So i have a plotWidget where i added multiple view-boxes that show different signals. Based on the multiplePlotAxes.py example i have created this example:

    import pyqtgraph as pg
    from pyqtgraph.Qt import QtCore, QtGui
    import numpy as np
    
    pg.mkQApp()
    
    pw = pg.PlotWidget()
    pw.show()
    pw.setWindowTitle('pyqtgraph example: MultiplePlotAxes')
    p1 = pw.plotItem
    p1.setMouseEnabled(x=True, y=False)
    p1.setLabels(left='axis 1')
    
    p2 = pg.ViewBox(name='axis 2')
    ax2 = pg.AxisItem('right')
    p1.layout.addItem(ax2, 2, 2)
    p1.scene().addItem(p2)
    ax2.linkToView(p2)
    p2.setXLink(p1)
    ax2.setLabel('axis 2', color='red')
    
    ## create third ViewBox. 
    ## this time we need to create a new axis as well.
    p3 = pg.ViewBox(name='axis 3')
    ax3 = pg.AxisItem('right')
    p1.layout.addItem(ax3, 2, 3)
    p1.scene().addItem(p3)
    ax3.linkToView(p3)
    p3.setXLink(p1)
    ax3.setLabel('axis 3', color='green')
    
    ## create third ViewBox.
    ## this time we need to create a new axis as well.
    p4 = pg.ViewBox(name='axis 4')
    ax4 = pg.AxisItem('right')
    p1.layout.addItem(ax4, 2, 4)
    p1.scene().addItem(p4)
    ax4.linkToView(p4)
    p4.setXLink(p1)
    ax4.setLabel('axis 4', color='blue')
    
    ## Handle view resizing 
    def updateViews():
        ## view has resized; update auxiliary views to match
        global p1, p2, p3, p4
    
        p2.setGeometry(p1.vb.sceneBoundingRect())
        p3.setGeometry(p1.vb.sceneBoundingRect())
        p4.setGeometry(p1.vb.sceneBoundingRect())

        p2.linkedViewChanged(p1.vb, p2.XAxis)
        p3.linkedViewChanged(p1.vb, p3.XAxis)
        p4.linkedViewChanged(p1.vb, p4.XAxis)

    def onSigRangeChanged(vb:pg.ViewBox):
        print(f"{vb.name} axis moved, i want it to be fully seen by expanding the view range")
        # [[xmin, xmax], [ymin, ymax]] = vb.viewRange()
        # print(ymin, ymax)

    updateViews()
    p1.vb.sigResized.connect(updateViews)
    p2.sigYRangeChanged.connect(onSigRangeChanged)
    p3.sigYRangeChanged.connect(onSigRangeChanged)
    p4.sigYRangeChanged.connect(onSigRangeChanged)

    p1.plot([1,2,4,8,16,32])
    p2.addItem(pg.PlotCurveItem([10,20,40,80,40,20], pen='r'))
    p3.addItem(pg.PlotCurveItem([123,456,789,987,654,321], pen='g'))
    p4.addItem(pg.PlotCurveItem([3200,1600,800,400,200,100], pen='b'))

    if __name__ == '__main__':
        pg.exec()
        
![](https://i.stack.imgur.com/aM2sg.png)

Basically when i grab the axis at (1) i want the main view-box to expand to cover the moved signal as well on Y axis.

X axis is locked for zooming.

I want the user to be able to "split" the signals but they should still be visible fully.

I tried working with sigYRangeChanged as seen in the example, but they Y range of the emitted viewBox does not show the range outside of the p1.vb, so i cannot resize to that one.

Any tips? Is there any magic i'm missing?

Or did i do it wrong? Do i need to add multiple axis-items in the p1 viewBox instead of adding multiple viewboxes?

**Comment by dziyaee** [comment]
I've struggled with something similar in the past. As long as the number of plots remains low and fixed, would you be fine just plotting them all on separate plot items and linking the x-axes?

Here is a minimal example including your 4 plots and data

```py
from pyqtgraph import GraphicsLayoutWidget, PlotItem, PlotDataItem, mkQApp


def updateViews(viewBox, viewRange):
    """Updates all other plot viewbox view xranges"""

    for plot in plots:
        if plot.vb is not viewBox:  # only update other ViewBoxes
            plot.vb.blockSignals(True)  # prevent this update from triggering other updates
            plot.setXRange(*viewRange, padding = 0)
            plot.vb.blockSignals(False)


# create app
app = mkQApp()

# create data items
item1 = PlotDataItem(y = [1, 2, 4, 8, 16, 32])
item2 = PlotDataItem(y = [10, 20, 40, 80, 40, 20], pen = 'r')
item3 = PlotDataItem(y = [123, 456, 789, 987, 654, 321], pen = 'g')
item4 = PlotDataItem(y = [3200, 1600, 800, 400, 200, 100], pen = 'b')

# create plots
plot1 = PlotItem()
plot2 = PlotItem()
plot3 = PlotItem()
plot4 = PlotItem()
plots = (plot1, plot2, plot3, plot4)

# only show bottom plot x-axis
plot1.getAxis("bottom").hide()
plot2.getAxis("bottom").hide()
plot3.getAxis("bottom").hide()

# link plot x-axes
plot1.vb.sigXRangeChanged.connect(updateViews)
plot2.vb.sigXRangeChanged.connect(updateViews)
plot3.vb.sigXRangeChanged.connect(updateViews)
plot4.vb.sigXRangeChanged.connect(updateViews)

# fix plot yaxis widths for all plots
plot1.getAxis("left").setWidth(50)
plot2.getAxis("left").setWidth(50)
plot3.getAxis("left").setWidth(50)
plot4.getAxis("left").setWidth(50)

# disable plot y-axis mouse wheel events
plot1.setMouseEnabled(x = True, y = False)
plot2.setMouseEnabled(x = True, y = False)
plot3.setMouseEnabled(x = True, y = False)
plot4.setMouseEnabled(x = True, y = False)

# add items to plots
plot1.addItem(item1)
plot2.addItem(item2)
plot3.addItem(item3)
plot4.addItem(item4)

# create window
window = GraphicsLayoutWidget()

# add plots to window
window.addItem(plot1, 0, 0)
window.addItem(plot2, 1, 0)
window.addItem(plot3, 2, 0)
window.addItem(plot4, 3, 0)

# run app
window.show()
app.exec()

```

You might even work in some sort of button or other control to switch between the split views and something like your original example where they were overlaid.

**Reply by mesa-git** [reply]
Hey,

Thanks for the reply. It looks like a decent work-around. I also have a InfiniteLine as a cursor that needs to go trough all plots which i have not tested on your example, but i think it should work.

I'm still trying to make it work with my original intention though.
There must be a way... some kind of signal that says if there is data not seen or somehow to use the autoPan function.

**Reply by dziyaee** [reply]
No problem! I'm definitely interested in a better solution as well. 

Have you tried seeing if you're able to recreate what you want just via method calls in interactive mode? What would your ideal plot look like?

**Reply by mesa-git** [reply]
_>method calls in interactive mode_
I'm not sure what that means.

At the moment i keep trying to look into the viewbox to somehow find what range would it need to display the signal fully...
And then trying to connect that to the main viewbox and expand it to show all signals.

I've looked into the viewbox https://pyqtgraph.readthedocs.io/en/latest/graphicsItems/viewbox.html

childrenTransform, childrenBounds, etc.

I would have assumed you can set the other viewBoxes to have the main one as parent? and then when you call autoRange
https://pyqtgraph.readthedocs.io/en/latest/_modules/pyqtgraph/graphicsItems/ViewBox/ViewBox.html#ViewBox.autoRange


_> What would your ideal plot look like?_

I have axis as groups of signals (because usually these groups have very difference scales for their values)
And i want to give the user the possibility to drag the axis to separate them and view them exactly like in your example, but instead of having 4 different plots above each other, have them plotted on the same graph, with the axis synced to 0, and if the user wants he can split them or not.

Cannot upload a image from work apparently :D to draw what i mean.

**Comment by dziyaee** [comment, marked as answer]
Hey again, not sure if I should have edited my first post but this solution is very different and hopefully closer to what you're looking for. I was inspired by your reply about having a main `ParentViewBox` and decided to see what I could do with that driving all other child `ViewBoxes`. Below is a minimal example containing two data curves. The difference in scale between them is x100.

Controls:
- mouse wheel over the `ParentViewBox` zooms all child `ViewBoxes`
- mouse drag over the `ParentViewBox` pans all child `ViewBoxes`
- mouse wheel over one of the yaxis `AxisItems` zooms only the corresponding `ViewBox`
- mouse drag over one of the yaxis `AxisItems` pans only the corresponding `ViewBox`
- Auto Button in the top left corner toggles autorange on all `ViewBoxes`

```py
import sys
import numpy as np
from PyQt5.QtWidgets import QApplication
from pyqtgraph import GraphicsLayoutWidget, ViewBox, AxisItem, PlotCurveItem, ButtonItem, icons


class ParentViewBox(ViewBox):
    """Forwards events to child viewboxes. Adds auto range button"""

    def __init__(self):
        super().__init__()
        # child viewboxes
        self._viewboxes = list()

        # autorange button (ripped out of PlotItem class)
        self.autoButton = ButtonItem(icons.getGraphPixmap('auto'), 14, self)
        self.autoButton.mode = 'auto'
        self.autoButton.clicked.connect(self.autoButtonClicked)

    def addViewBox(self, viewbox):
        """add a child viewbox"""

        self._viewboxes.append(viewbox)

    def resizeEvent(self, event):
        """resize all viewboxes to match parent"""

        super().resizeEvent(event)
        rect = self.sceneBoundingRect()
        for viewbox in self._viewboxes:
            viewbox.setGeometry(rect)

    def wheelEvent(self, event, axis = None):
        """zoom all viewboxes if parent zoomed"""

        for viewbox in self._viewboxes:
            viewbox.wheelEvent(event, axis)

    def mouseDragEvent(self, event, axis = None):
        """drag all viewboxes if parent dragged"""

        for viewbox in self._viewboxes:
            viewbox.mouseDragEvent(event, axis)

    def autoButtonClicked(self):
        """toggle autorange on all viewboxes if button clicked"""

        if self.autoButton.mode == 'auto':
            for viewbox in self._viewboxes:
                viewbox.enableAutoRange()

        else:
            for viewbox in self._viewboxes:
                viewbox.disableAutoRange()


class XAxis(AxisItem):
    """Disables mouse drag interaction"""

    def __init__(self):
        super().__init__("bottom")

    def mouseDragEvent(self, event):
        """Ensure that user can't drag curves out of alignment by dragging x-axis"""
        pass


# app
app = QApplication(sys.argv)
window = GraphicsLayoutWidget()
window.show()

# curves
data1 = np.random.randn(100)
data2 = np.random.randn(100) * 100

curve1 = PlotCurveItem(y = data1, pen = "r", name = "curve1")
curve2 = PlotCurveItem(y = data2, pen = "g", name = "curve2")

# child view for each curve
view1 = ViewBox()
view2 = ViewBox()

view1.addItem(curve1)
view2.addItem(curve2)

window.scene().addItem(view1)
window.scene().addItem(view2)

# parent view contains all child views and drives mouse interaction on the viewbox
parentView = ParentViewBox()
parentView.addViewBox(view1)
parentView.addViewBox(view2)
window.addItem(parentView, 0, 0)

# yaxis for each curve / child view
yaxis1 = AxisItem("right")
yaxis2 = AxisItem("right")

yaxis1.setLabel(text = curve1.name(), color = curve1.opts["pen"].color().name())
yaxis2.setLabel(text = curve2.name(), color = curve2.opts["pen"].color().name())

yaxis1.linkToView(view1)
yaxis2.linkToView(view2)

window.addItem(yaxis1, 0, 1)
window.addItem(yaxis2, 0, 2)

# one xaxis linked to curve1 (for display purposes only)
xaxis = XAxis()
xaxis.linkToView(view1)
window.addItem(xaxis, 1, 0)


if __name__ == "__main__":
    # handle interactive mode
    if sys.flags.interactive == 0:
        sys.exit(app.exec())
```

Demo:

https://user-images.githubusercontent.com/13340419/167010736-d5f9dd66-baee-4715-9a53-e60795b68a22.mp4

I'm not sure if there are any side effects to forwarding mouse events to multiple objects to process, so just keep that in mind.

**Reply by mesa-git** [reply]
Hey, thanks for the reply.

This is different from what i imagined but does the job almost exactly the same way. I will try it to see how it feels.

**Comment by mesa-git** [comment]
I ended up using a combination of stuff that were suggested here...
I have a plotItem that I use which has the main "axis"...

```
        self.addNewAxis(AXIS_FUNC)
        self.addNewAxis(AXIS_LABELS)
```

I add new axis using this:

```
def addNewAxis(self, name_of_axis):
        newVb = ViewBox()
        self.viewAxis[name_of_axis] = newVb
        newAxis = AxisItem('right')
        self.plotInstance.layout.addItem(newAxis, 2, len(self.viewAxis))
        self.plotInstance.scene().addItem(newVb)
        newAxis.linkToView(newVb)
        newVb.setXLink(self.plotInstance)
        labelStyle = {'color': HEX_WHITE, 'font-size': '10pt'}
        newAxis.setLabel(name_of_axis, **labelStyle)
        self.updateViews()

```

This somehow throws and warning:

```
QGridLayoutEngine::addItem: Cell (2, 2) already taken
QGridLayoutEngine::addItem: Cell (2, 2) already taken
```

but works... and I add signals from an internal data structure using:

```
            if data['axis'] == AXIS_MAIN:
                plotted_signal = self.plotInstance.plot(*plot_signal_data,
                                                        stepMode=step,
                                                        symbol=symbol,
                                                        symbolSize=symbolSize,
                                                        symbolPen=symbolPen,
                                                        symbolBrush=symbolBrush,
                                                        pen=mkPen(color=data['color'], width=sigWidth),
                                                        skipFiniteCheck=True)
            else:
                plotted_signal = PlotDataItem(*plot_signal_data,
                                               stepMode=step,
                                               symbol=symbol,
                                               symbolSize=symbolSize,
                                               symbolPen=symbolPen,
                                               symbolBrush=symbolBrush,
                                               pen=mkPen(color=data['color'], width=sigWidth),
                                               skipFiniteCheck=True)
                self.viewAxis[data['axis']].addItem(plotted_signal)
            self.legend.addItem(plotted_signal, signal)
```
